src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -414,6 +414,11 @@ BlockIO InterpreterInsertQuery::execute()
         res.out = std::move(out_streams.at(0));
 
     res.pipeline.addStorageHolder(table);
+    if (const auto * mv = dynamic_cast<const StorageMaterializedView *>(table.get()))
```

Comment:
When we add one storage holder, do we always have to add the entire hierarchy? 

There is also a similar pipeline construction in `QueryPlan/ReadFromStorage.cpp:94`, does it need any treatment?

Maybe we have to adopt some simple ownership strategy, like 'a storage holder also holds all nested storages'...

Reply:
`ReadFromStorageStep` doesn't know about nested storages, so if some storage calls `read` or `write` for another one, then it's responsible for table locks and lifetime of nested storages. However, in `write` method we cannot add storage holders to pipeline, because `IStorage::write(...)` returns `BlockOutputStreamPtr`, so when inserting directly into MV we add it here.